jvb/src/main/kotlin/org/jitsi/videobridge/relay/Relay.kt

```diff
@@ -0,0 +1,629 @@
+import org.jitsi.xmpp.extensions.colibri.WebSocketPacketExtension
+import org.jitsi.xmpp.extensions.jingle.DtlsFingerprintPacketExtension
+import org.jitsi.xmpp.extensions.jingle.IceUdpTransportPacketExtension
+import org.json.simple.JSONObject
+import java.time.Clock
+import java.time.Instant
+import java.util.concurrent.TimeUnit
+import java.util.function.Supplier
+
+/**
+ * Models a relay (remote videobridge) in a [Conference].
+ */
+/* TODO: figure out how best to share code between this and [Endpoint], without multiple inheritance. */
+class Relay @JvmOverloads constructor(
+    /**
+     * The unique identifier of this [Relay]
+     */
+    val id: String,
+    /**
+     * The [Conference] this [Relay] belongs to.
+     */
+    val conference: Conference,
+    parentLogger: Logger,
+    /**
+     * True if the ICE agent for this [Relay] will be initialized to serve as a controlling ICE agent, false otherwise.
+     */
+    iceControlling: Boolean,
+    useUniquePort: Boolean,
+    clock: Clock = Clock.systemUTC()
+) : EncodingsManager.EncodingsUpdateListener, PotentialPacketHandler {
+
+    private val eventEmitter: EventEmitter<AbstractEndpoint.EventHandler> = SyncEventEmitter()
+
+    /**
+     * The [Logger] used by the [Relay] class to print debug information.
+     */
+    private val logger = createChildLogger(parentLogger).apply { addContext("relayId", id) }
+
+    /**
+     * The indicator which determines whether [expire] has been called on this [Relay].
+     */
+    private var expired = false
+
+    private val iceTransport = IceTransport(id, iceControlling, useUniquePort, logger, clock)
+    private val dtlsTransport = DtlsTransport(logger)
+
+    private val diagnosticContext = conference.newDiagnosticContext().apply {
+        put("relay_id", id)
+    }
+
+    private val timelineLogger = logger.createChildLogger("timeline.${this.javaClass.name}")
+
+    private val relayedEndpoints = HashMap<String, RelayedEndpoint>()
+    private val endpointsBySsrc = HashMap<Long, RelayedEndpoint>()
+    private val endpointsLock = Any()
+
+    /**
+     * Listen for RTT updates from [transceiver] and update the ICE stats the first time an RTT is available. Note that
+     * the RTT is measured via RTCP, since we don't expose response time for STUN requests.
+     */
+    private val rttListener: EndpointConnectionStats.EndpointConnectionStatsListener =
+        object : EndpointConnectionStats.EndpointConnectionStatsListener {
+            override fun onRttUpdate(newRttMs: Double) {
+                if (newRttMs > 0) {
+                    transceiver.removeEndpointConnectionStatsListener(this)
+                    iceTransport.updateStatsOnInitialRtt(newRttMs)
+                }
+            }
+        }
+
+    /* TODO: we eventually want a smarter Transceiver implementation, that splits processing by
+     *  source endpoint or something similar, but for the initial implementation this should work.
```

Comment:
So this loses the concept of demuxing remote incoming packets to individual receive pipelines per remote endpoint?  Before I think I had it set up where there were independent rx pipelines leveraging a shared transport--but maybe I'm misremembering.

Reply:
Yeah - I have this as a TODO in case it ends up being useful, but it'd be pretty complicated, because (unlike old Octo) this has NACK support, so we'd need to wire up retransmission across receive pipelines (because you can't tell which remote source a NACK is for until the SRTCP has been decrypted). We'd also need to propagate SRTP keys across transceivers, etc.  None of this is impossible, but I thought it wasn't needed for MVP.

We do have separate receive and send pipelines per remote relay, so once you get more than two bridges in a conference, we get some degree of parallelization.

I haven't done the performance testing of this yet; whether there's any bottleneck here is one of the things I'll be looking for.

Reply:
I remember refactoring the Octo stuff we had to move _away_ from the single collection of endpoints and model them more like normal endpoints, just with the shared transport, etc. stuff.  Performance I do think was one benefit, but I think what I liked most is that it felt like it fit more cleanly into the design without these being special-cased.  I think for the most part that worked?

I think I'm missing something about the nack scenario you're talking about: are you talking about the JVB nacking a packet that wasn't received from a remote endpoint?  Or a nack sent to the JVB from a local endpoint that corresponds to a remote endpoint?  I guess not the former, because that seems straightforward.  Maybe not the latter either, because we already have to decrypt all RTCP coming in from local endpoints anyways, no?

Reply:
I'm talking about a JVB NACKing a packet that wasn't received from the remote JVB - the purpose of this project was to allow Octo connections to go over the public Internet, and one of the things that means is that this connection might be losing packets.

If you have a `Transceiver` object per remote endpoint, then when the upstream JVB receives the NACK, it doesn't know which `Transceiver` it should feed the NACK to.  We'd have to wire up some sort of cross-`Transceiver` NACK handling in order to get this to work.

Reply:
Ah ha, I see now.  It's a common set of SRTP keys for all the end points, right?  Maybe this would involve moving the SRTP decrypt into the 'common' logic shared across all the receivers, and the receive pipelines pick up after that.  This would also mean they'd have separate entry points for RTP and RTCP (since that demuxing would've already been done).  Given the SRTP is probably on the heavier side of the work, it does mean the performance gain of parallelization would be lost, but maybe that's just how it would have to be.  Would keep the relayed endpoints less "special" potentially, though.